File: javascript/anagram/anagram.js
alphagram = function(word) {
  return word.split('').sort().join('');
};

are_anagrams = function(l, r){
  left = l.toLowerCase();
  right = r.toLowerCase();

  if (left == right ) return false;

  return alphagram(left) == alphagram(right);
};

Anagram = function(word){
  this.word = word;
};
Anagram.prototype.match = function(words){
  return words.filter(function(word){
    return are_anagrams(this.word, word);
  },this);
};

module.exports = Anagram;

File: ruby/anagram/anagram.rb
class Anagram
  def initialize(word)
    @word = word.downcase
  end

  def anagram_of?(other)
    self.letters == other.letters &&
       self.word != other.word
  end

  def match words
    words.select do |word|
      anagram_of?(Anagram.new(word))
    end
  end

  protected

  attr_reader :word

  def letters
    @letters ||= word.chars.sort
  end
end
